akka-typed/src/main/scala/akka/typed/internal/ActorCell.scala

```diff
@@ -0,0 +1,433 @@
+  val parent: ActorRefImpl[Nothing])
+    extends ActorContext[T] with Runnable with SupervisionMechanics[T] with DeathWatch[T] {
+  import ActorCell._
+
+  /*
+   * Implementation of the ActorContext trait.
+   */
+
+  protected var childrenMap = Map.empty[String, ActorRefImpl[Nothing]]
+  protected var terminatingMap = Map.empty[String, ActorRefImpl[Nothing]]
+  override def children: Iterable[ActorRef[Nothing]] = childrenMap.values
+  override def child(name: String): Option[ActorRef[Nothing]] = childrenMap.get(name)
+  protected def removeChild(actor: ActorRefImpl[Nothing]): Unit = {
+    val n = actor.path.name
+    childrenMap.get(n) match {
+      case Some(`actor`) ⇒ childrenMap -= n
+      case _ ⇒
+        terminatingMap.get(n) match {
+          case Some(`actor`) ⇒ terminatingMap -= n
+          case _             ⇒
+        }
+    }
+  }
+  private[typed] def terminating: Iterable[ActorRef[Nothing]] = terminatingMap.values
+
+  private var _self: ActorRefImpl[T] = _
+  private[typed] def setSelf(ref: ActorRefImpl[T]): Unit = _self = ref
+  override def self: ActorRefImpl[T] = _self
+
+  protected def ctx: ActorContext[T] = this
+
+  override def spawn[U](props: Props[U], name: String): ActorRef[U] = {
+    if (childrenMap contains name) throw new InvalidActorNameException(s"actor name [$name] is not unique")
+    if (terminatingMap contains name) throw new InvalidActorNameException(s"actor name [$name] is not yet free")
+    val cell = new ActorCell[U](system, props, self)
+    val ref = new LocalActorRef[U](self.path / name, cell)
+    cell.setSelf(ref)
+    childrenMap = childrenMap.updated(name, ref)
+    ref.sendSystem(Create())
+    ref
+  }
+
+  private var nextName = 0L
+  override def spawnAnonymous[U](props: Props[U]): ActorRef[U] = {
+    val name = Helpers.base64(nextName)
+    nextName += 1
+    spawn(props, name)
+  }
+
+  override def stop(child: ActorRef[Nothing]): Boolean = {
+    val name = child.path.name
+    childrenMap.get(name) match {
+      case None                      ⇒ false
+      case Some(ref) if ref != child ⇒ false
+      case Some(ref) ⇒
+        ref.sendSystem(Terminate())
+        childrenMap -= name
+        terminatingMap = terminatingMap.updated(name, ref)
+        true
+    }
+  }
+
+  protected def stopAll(): Unit = {
+    childrenMap.valuesIterator.foreach { ref ⇒
+      ref.sendSystem(Terminate())
+      terminatingMap = terminatingMap.updated(ref.path.name, ref)
+    }
+    childrenMap = Map.empty
+  }
+
+  override def schedule[U](delay: FiniteDuration, target: ActorRef[U], msg: U): Cancellable =
+    system.scheduler.scheduleOnce(delay)(target ! msg)(ExecutionContexts.sameThreadExecutionContext)
+
+  override val executionContext: ExecutionContextExecutor = system.dispatchers.lookup(props.dispatcher)
+
+  override def spawnAdapter[U](f: U ⇒ T): ActorRef[U] = {
+    val name = Helpers.base64(nextName)
+    nextName += 1
+    val ref = new FunctionRef[U](self.path / name,
```

Comment:
Okay, convinced ;-) I made the prefix for adapters "$!" so that they can be recognized. But using a different counter seems a waste. If optimizations become necessary, we can do them later.